query-planner-js/src/buildQueryPlan.ts

```diff
@@ -742,7 +742,21 @@ function completeField(
     splitSubfields(context, fieldPath, subfields, subGroup);
     debug.groupEnd();
 
-    parentGroup.otherDependentGroups.push(...subGroup.dependentGroups);
+    // We need to hoist dependent groups of the subgroup to the parent group.
+    // In order to avoid duplicate fetches, we try to find existing dependent
+    // groups with the same service and merge path first.
+    for (const dependentGroup of subGroup.dependentGroups) {
+      const existingDependentGroup = parentGroup.otherDependentGroups.find(
```

Comment:
Hmmm, should this be `parentGroup.dependentGroups` (which refers to both `dependentGroupsByService` and `otherDependentGroups`)?

Reply:
@martijnwalraven Random drive-by thought. When I recently reviewed the query planner, I found the aspect of FetchGroup where there are two separate sets of subgroups to be pretty confusing.  When I dived in farther, it seemed like `dependentGroupsByService` was only really written from inside `splitSubfields`, which is called only from `completeField` with the FetchGroup `subGroup`, and `subGroup` isn't added directly to the tree itself but instead its dependent groups get extracted and added to the `otherDependentGroups` of an existing tree member.

So it seemed to me there there might be a simplifying refactor where `FetchGroup` just has a set/list of dependent groups (like `otherDependentGroups` now), and the logic currently done in `dependentGroupForService` could be specific to `splitSubfields` or `completeField`.  ie, `dependentGroupsByService` could perhaps be a local variable used in building up the FetchGroup rather than hardcoded into the `FetchGroup` structure.

I tried doing this refactor for a few minutes and it wasn't 100% trivial (eg, the `baseGroup.dependentGroupForService` call is a bit different), but it might be feasible and increase clarity?

(Alternatively, maybe the opposite refactor where `dependentGroupsByService` is the only structure that needs exist could happen, and the single write to `otherDependentGroups` in `completeSubfield` could instead somehow merge `subGroups.dependentGroups` into `parentGroup.dependentGroupsByService`?)

Reply:
Yeah, agreed this is confusing, and I think refactoring it would help. I wanted to keep this fix as small as possible, especially since we plan on making larger changes to the query planner soon.